### PR TITLE
Martin/ios 315 prepare to send for threads

### PR DESCRIPTION
**Pull request by MartinJibber**
DropZone and send preparation logic is now in a common delegate object.  The Conversation List VC and Thread VC both use this now.

Cell Reuse behavior is now improved so cells don't lose their send state when recycled.

The Thread VC message prep now works properly.

https://user-images.githubusercontent.com/88675954/145097430-1baae824-ed98-459e-a37f-3477c911022f.mp4

### CODE REVIEW

File: Shared/Models/Message Model/MessageSequence.swift [hunk 1]
```diff
@@ -11,19 +11,20 @@ import Foundation
...
-    var conversationId: String { get }
     var createdAt: Date { get }
     var updatedAt: Date { get }
     var isCreatedByCurrentUser: Bool { get }
     var authorID: String { get }
     var attributes: [String: Any]? { get }
     var messages: [Messageable] { get }
+
+    // Stream variables
+    var streamCID: ConversationID? { get }
```
Comment (MartinJibber) [inline Shared/Models/Message Model/MessageSequence.swift:22]: Converting from string to ConversationID is really inconvenient.

File: Shared/Models/Message Model/MessageSequence.swift [hunk 2]
```diff
@@ -11,19 +11,20 @@ import Foundation
...
 }
 
 func ==(lhs: MessageSequence, rhs: MessageSequence) -> Bool {
     guard type(of: lhs) == type(of: rhs) else { return false }
     return lhs.id == rhs.id
-        && lhs.conversationId == rhs.conversationId
```
Comment (MartinJibber) [inline Shared/Models/Message Model/MessageSequence.swift:26]: The ID is all the matters for uniqueness.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 3]
```diff
@@ -72,7 +72,15 @@ class ConversationMessagesCell: UICollectionViewCell {
...
-        self.dataSource.update(messageSequence: sequence)
+        self.dataSource.set(messageSequence: sequence)
+    }
+
+    func set(layoutForDropZone: Bool) {
+        self.collectionLayout.layoutForDropZone = layoutForDropZone
+    }
+
+    func set(isPreparedToSend: Bool) {
+        self.dataSource.shouldPrepareToSend = isPreparedToSend
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:83]: To improve cell reuse behavior, set a persistent variable on the datasource and layout objects for send preparation.

Now we won't lose the preparation UI when scrolling around.

File: iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift [hunk 4]
```diff
@@ -130,20 +138,6 @@ class ConversationMessagesCell: UICollectionViewCell {
...
-        guard let conversation = self.conversation else { return }
-        self.set(sequence: conversation)
-    }
-
-    func unprepareForNewMessage(reloadMessages: Bool) {
-        self.dataSource.prepareForSend = false
-
-        guard reloadMessages, let conversation = self.conversation else { return }
-        self.set(sequence: conversation)
-    }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/ConversationMessagesCell.swift:146]: Replaced by the code above.

File: iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift [hunk 5]
```diff
@@ -21,6 +21,7 @@ class MessagesTimeMachineCollectionViewLayout: TimeMachineCollectionViewLayout {
 
     // MARK: - Layout Configuration
 
+    var layoutForDropZone: Bool = false
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift:24]: This is a message specific feature so I moved it out of the base class.

File: iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift [hunk 6]
```diff
@@ -120,7 +121,6 @@ class MessagesTimeMachineCollectionViewLayout: TimeMachineCollectionViewLayout {
     /// If true, scroll to the most recent item after performing collection view updates.
     private var shouldScrollToEnd = false
     private var deletedIndexPaths: Set<IndexPath> = []
-    private var insertedIndexPaths: Set<IndexPath> = []
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift:123]: For now, I'm going to let messages fade in when sent. I made the message preview view fade out so it feels smoother.

File: iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift [hunk 7]
```diff
@@ -169,27 +167,11 @@ class MessagesTimeMachineCollectionViewLayout: TimeMachineCollectionViewLayout {
 
     override func finalLayoutAttributesForDisappearingItem(at itemIndexPath: IndexPath)
     -> UICollectionViewLayoutAttributes? {
-        guard deletedIndexPaths.contains(itemIndexPath) else { return nil }
+        // Items that are just moving are marked as "deleted" by the collection view.
+        // Only animate changes to items that are actually being deleted otherwise weird animation issues
+        // will arise.
+        guard self.deletedIndexPaths.contains(itemIndexPath) else { return nil }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift:173]: Adding explanations for this weird code for posterity.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 8]
```diff
@@ -27,12 +27,10 @@ enum ConversationUIState {
 class ConversationListViewController: FullScreenViewController,
                                       UICollectionViewDelegate,
                                       UICollectionViewDelegateFlowLayout,
-                                      SwipeableInputAccessoryViewDelegate, ActiveConversationable {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:30]: The Convo List and Thread VCs are no longer direct delegates of the swipe accessory. Instead there is a separate delegate object.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 9]
```diff
@@ -46,10 +44,13 @@ class ConversationListViewController: FullScreenViewController,
...
     lazy var messageInputAccessoryView: ConversationInputAccessoryView = {
         let inputView: ConversationInputAccessoryView = ConversationInputAccessoryView.fromNib()
-        inputView.delegate = self
+        inputView.delegate = self.swipeInputDelegate
         inputView.textView.restorationIdentifier = "list"
         return inputView
     }()
+    lazy var swipeInputDelegate = SwipeableInputAccessoryMessageSender(viewController: self,
+                                                                       collectionView: self.collectionView,
+                                                                       isConversationList: true)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:53]: Here's the delegate that handles showing the drop zone and adjusting the collection view offsets, etc.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 10]
```diff
@@ -310,190 +310,31 @@ class ConversationListViewController: FullScreenViewController,
     func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
         self.updateCenterMostCell()
     }
+}
 
-    // MARK: - SwipeableInputAccessoryViewDelegate
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:314]: All of the following is moved into the delegate object.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 11]
```diff
@@ -310,190 +310,31 @@ class ConversationListViewController: FullScreenViewController,
...
-            }
-
-            UIView.animate(withDuration: Theme.animationDurationStandard) {
-                self.sendMessageDropZone.setState(.newMessage, messageColor: self.collectionView.getDropZoneColor())
-            }
+// MARK: - MessageSendingViewControllerType
 
-            return false
-        }
+extension ConversationListViewController: MessageSendingViewControllerType {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:317]: The delegate object requires that VC implement these helper functions.

File: iOS/Flows/Conversation/SwipeableInputAccessoryMessageSender.swift [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+    func set(messageSequencePreparingToSend: MessageSequence?, reloadData: Bool)
+    func sendMessage(_ message: Sendable)
+    func createNewConversation(_ sendable: Sendable)
+}
+
+protocol MessageSendingCollectionViewType: CollectionView {
+    func getMessageDropZoneFrame(convertedTo view: UIView) -> CGRect
+    func getDropZoneColor() -> Color?
+    func getNewConversationContentOffset() -> CGPoint
+}
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/SwipeableInputAccessoryMessageSender.swift:23]: Any VC that wants to use this delegate object needs to conform to these protocols.

File: iOS/Flows/Conversation/SwipeableInputAccessoryMessageSender.swift [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        self.isConversationList = isConversationList
+    }
+
+    /// The collection view's content offset at the first call to prepare for a swipe. Used to reset the the content offset after a swipe is cancelled.
+    private var initialContentOffset: CGPoint?
+    /// The last swipe position type that was registersed, if any.
+    private var currentSendMode: SendMode?
+
+    // MARK: - SwipeableInputAccessoryViewDelegate
+
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/SwipeableInputAccessoryMessageSender.swift:61]: All logic is no longer duplicated between the thread and conversation list VCs. It's been generalized to work with both.

File: iOS/Flows/Conversation/Threads/ThreadCollectionView.swift [hunk 14]
```diff
@@ -10,7 +10,7 @@ import Foundation
 
 class ThreadCollectionView: CollectionView {
 
-    let threadLayout = ThreadCollectionViewLayout()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Threads/ThreadCollectionView.swift:13]: Threads don't need their own layout subclass. Deleting it.

File: iOS/Flows/Conversation/Threads/ThreadViewController.swift [hunk 15]
```diff
@@ -169,127 +171,44 @@ class ThreadViewController: DiffableCollectionViewController<MessageSequenceSect
...
-        var cycle = super.getAnimationCycle()
-        cycle?.shouldConcatenate = false
-        // Scroll to the lastest reply.
-        if let threadLayout = self.collectionView.collectionViewLayout as? ThreadCollectionViewLayout {
-            let lastReplyIndex = clamp(self.messageController.replies.count - 1, min: 0)
-            let yOffset = CGFloat(lastReplyIndex) * threadLayout.itemHeight
-            cycle?.scrollToOffset = CGPoint(x: 0, y: yOffset)
-        }
-
-        return cycle
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Threads/ThreadViewController.swift:181]: The collection view layout handles scrolling to the latest reply now.

File: iOS/Flows/Conversation/Threads/ThreadViewController.swift [hunk 16]
```diff
@@ -169,127 +171,44 @@ class ThreadViewController: DiffableCollectionViewController<MessageSequenceSect
...
+        return nil
     }
+}
 
-    // MARK: - SwipeableInputAccessoryViewDelegate
-
-    var prepareForSend = false
-
-    func swipeableInputAccessory(_ view: SwipeableInputAccessoryView, swipeIsEnabled isEnabled: Bool) {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Threads/ThreadViewController.swift:188]: Similar to the ConvoList VC, all the red here has been moved to the delegate object.

File: iOS/Flows/People/PeopleCoordinator+Extensions.swift [hunk 17]
```diff
@@ -73,18 +73,18 @@ extension PeopleCoordinator {
             async let matchingUser = User.getFirstObject(where: "phoneNumber", contains: phone)
             // Ensure that the reservation metadata is prepared before we show the reservation
             try await reservation.prepareMetadata(andUpdate: [])
-            reservation.conversationId = self.conversationId
+            reservation.conversationId = self.conversationID?.description ?? String()
```
Comment (MartinJibber) [inline iOS/Flows/People/PeopleCoordinator+Extensions.swift:76]: This is a consequence of making conversationID an actual ChannelID type.

File: iOS/Collection View Management /CollectionViewDataSource.swift [hunk 18]
```diff
@@ -179,6 +179,13 @@ extension CollectionViewDataSource {
         }
     }
 
+    func reconfigureAllItems() {
+        self.applyChanges{ snapshot in
+            let allItems = snapshot.itemIdentifiers
+            snapshot.reconfigureItems(allItems)
+        }
+    }
+
```
Comment (MartinJibber) [inline iOS/Collection View Management /CollectionViewDataSource.swift:188]: Convenience function.

**Review by Benjamindodgson — APPROVED** [review]

File: iOS/Collection View Management /CollectionViewDataSource.swift (hunk 18 above)
Comment (Benjamindodgson) [reply iOS/Collection View Management /CollectionViewDataSource.swift:188]: me likey

**Pull request merged**